### DISCUSSION TITLE
Diffusion Model Details

### DISCUSSION THREAD

**Discussion by Malik7115** [neonbjb/tortoise-tts#134] [Q&A]
Hi, I am back! 

after spending a lot of time on diffusion models I tried to do replicate your experiment, with the original repo by lucidrains. I dont have immense data, so trying with a limited number of speakers. I conditioned my vqvae codes onto the unet. Although the speech had some human form, but was completely giberish in terms of language. I noticed you alleviated this problem by using a surrogate loss as mentioned here: https://nonint.com/2022/04/04/190/. I am a bit confused, are you using a simple network to get a loss between the embeddings of the codes and the codes themselves?

P.S. I have opened a new discussion panel, so that its easier for people to navigate and ask questions.

**Comment by neonbjb** [comment]
Hey, glad to talk to you again!

Which diffusion model did you end up attempting to train? The problem you mention (where a surrogate loss was required) was unique to the U-net architecture. After I moved over to the "flat diffusion" models (which are just transformer encoders trained with diffusion objectives), I no longer needed the surrogate losses.

[This](https://github.com/neonbjb/DL-Art-School/blob/master/codes/models/audio/tts/unet_diffusion_tts_flat.py) is the model architecture used in Tortoise. Though, if you're feeling a bit brave, I would highly recommend using [this](https://github.com/neonbjb/DL-Art-School/blob/master/codes/models/audio/music/transformer_diffusion12.py#L332) architecture instead. It uses something similar to RRDB with some attention mixed in and drops out the conditioning input, which I found to not be useful in the diffusion decoder. I have trained one of these models on the Tortoise codes and the performance (by tts-scores) is superior to what shipped with this repo.

**Reply by Malik7115** [reply]
Well I first when with the excellent DDPM repo by lucidrains https://github.com/lucidrains/denoising-diffusion-pytorch. However, it did not contain the functionality of improved diffusion so I went for the native implementation of the improved diffusion repo by OpenAI (I saw that you mentioned something in your tortoise Repo). This repo had the functionality of sampling with fewer iterations but the Unet was a big problem. Although, I could get the speaker style conditioning (which I am doing in a supervised fashion since I have limited data) but the speech was utter gibberish, even after many many iterations. 

I'm curious though, isnt the Unet a much deeper and complex network, why does it not train?

**Reply by neonbjb** [reply]
You may already know this, but it's worth saying just in case you don't: The model architecture of a diffusion model can be totally dissociated from the diffusion algorithm. What I mean by that is that you (should be able to) use any Unet model (or my flat diffusion models) with either the lucidrains diffusion implementation or the OpenAI one (or the diffusers library from huggingface!)

Diffusion models also train really, really slowly and things like coherent words will be one of the later things to emerge while training the model. Do you have any loss curves you can share with me? How many samples did you expose your model to?

Ensuring your alignment is correct (between the codes you provide and the target audio) is also something I would pay close attention to. Any bugs here will induce the model to ignore your codes and produce unconditional outputs, which will sound like gibberish like you describe. There will be almost no difference in losses between an unconditioned model and a conditioned one.

> I'm curious though, isnt the Unet a much deeper and complex network, why does it not train?

It is more complex, but that complexity is focused in the lower layers of the network. If you want to really blow your mind, print out the parameter density of each layer of your Unet. In the models I trained, something like 90% of the parameters were in the bottom few layers. These layers are the last to get trained (they can't stabilize until the top layers do), and they get really noisy gradient signals. It is almost certainly possible to train an effective Unet model, but flat architectures are less errorprone.

**Reply by Malik7115** [reply]
Thanks for the detailed reply. I am still a novice in this field and trying my best to learn :) 
TBH I went guns blazing while training for my first iteration and did not save any loss logs as graphs. I only have a log file that i referenced to check the overall performance of the model. As for samples, I think i exposed the model to about 50 hrs of data.

**Reply by neonbjb** [reply]
Ah OK. I would highly recommend building an evaluation harness around your diffusion model to track training progress. Every 1000 iterations or so you can perform the full diffusion process on 100 samples or so and run them through a wav2vec2 model (or your CLVP model).

This will let you track whether or not you are making forward progress with the training. If not, something is wrong and you should debug. This is largely how I was able to ultimately finish up Tortoise.

**Reply by Malik7115** [reply]
Hello Hello, 
I am so excited to inform you that the diffusion is now finally speaking....with my limited dataset it is no longer gibberish anymore. The quality is not exactly what I would call great but I think I will stop it here and move towards the AR model now. Ill open another discussion panel once I start on it, so that it might be helpful to others as well.

**Reply by neonbjb** [reply]
Awesome! Glad to hear that! 👍👍👍

What ended up being the problem?

**Reply by Malik7115** [reply]
I am pretty sure it was the Unet...;)
